### PR TITLE
Collection tree

### PR DESCRIPTION
**Pull request by dicksonlaw583**
This pull request adds the collection tree feature to the manifest selector. Updated documentation and tests are included with the feature.

### CODE REVIEW

**Comment by mejackreed** [comment]
Thanks so much for putting this into a PR! 

Is it possible to move to the pattern of including jstree as an npm dependency instead of committing the library within the repository?

This could remove us from having to maintain the following:

 - `css/jstree.css`
 - `images/jstree/32px.png`
 - `images/jstree/40px.png`
 - `images/jstree/throbber.gif`
 - `js/lib/jstree.min.js`

File: package.json [hunk 1]
```diff
@@ -47,7 +47,7 @@
     "sinon": "^1.17.6"
   },
   "scripts": {
-    "start": "./node_modules/.bin/grunt serve",
+    "start": "node node_modules\/.bin\/grunt serve",
```
Comment (csbailey5t) [inline package.json]: Is this change necessary?

Comment (dicksonlaw583) [inline package.json]: It isn't, but I have no idea how that made its way in there.

**Comment by csbailey5t** [comment]
At the broad level, it would be nice to move feature specific styles into an appropriately named .less file, within the `mirador/css/less` dir, given the refactor to less earlier this year.

**Comment by rsinghal** [comment]
Agreed, this PR should not add mirador.css back into the code base, as it was removed for the .less files

**Comment by dicksonlaw583** [comment]
The collection tree style has been moved into LESS.
 
Unfortunately, the NPM version of jstree has the icon images mixed into the CSS directory (similar issue with qtip), so that part remains separate with a manually edited css/jstree.css.

File: package.json (hunk 1 above)
Comment (dicksonlaw583) [inline package.json]: The change has now been undone. I will watch out for this in the future.

File: css/jstree.css [hunk 2]
```diff
@@ -400,9 +400,23 @@
...
...VOnTpV3p4A0epJlyDA5RRIgAQlQNHGAOcP/dXPdoXWRGczRGV5ZTfjUC9MywnBxAyGiWkME9IY7tsepmEvDBsr6htsLcBeveZ9ddddTRWMoYox+JOghTE0P/IIjEuWNLa+RKb7CXCaDG6PC/+sPCABEpQARZ0Flg/77f07pXWPpjCMS2YYq2QYyzGMSWIYncQw5L1e3zDZ8ImiG9pXSsbkJ3r3btEwhlLGUO6bFKllDAbG0MAYHBMmoH7hwpaDQ4bIO0qAHXkKPH3GdLnH7RH8ic+P94+2wZAESFACFPsyGPmQpwa9K619eB/DqD0+FAz3vh1WymRDxwflDTLGkkNC5h6WSIQzjOE8Y7jKGK4xhmu9e+O6RAK9b3LEEBaGm+PHg587Vzg8ePCc7jYJMmPGDLnb7XbCDbhdLrgFwYvLBZfggksQAhmQBEhQAgzGdYDzhz7747d71Y78WIoRH0lxz/qwa2zufVFBMzRjTBkW9te8gQPxRUgIihnDRcZwPiwMB/r2PZjC2KErjKHalwjr+/cHHx2N/GHDNgVTgB0tv+joaMn06dPlzpaWFpcgQBCccDq9uJwCXE6nd/xPEOChBEhQAgzihdCyob96ZOtI06h3RpbdKfmJKEBJGmPDEsPC/nF44EAUhYbiZEgIUqTS9DTGHkxj7EElY+kXfOOCX991F3LvuecfaYwN624XQk+fPl1ua7I57TY77HYftttxCQIlQIISYBBXgkiYbMh/tZ7tDWYCZIxJ9jN2f1JYWELmgAFQMnbAd6mLf6XI8CTGDpxgDJkSScJ+xu4PymUwnbwU7rHHHpNZrBZXY1Mj2nCzEdbGRlibrLA2WmG32SkBEpQAu9nNECS+tPdiGmMRjDHJLa9F+F570P9ad7sZwtSpU2VOZ4u7ubkFDocDdpsNDrsDzY5mCC1OCIITLsFJkyAEJcBuejssSSv+6Wvd7csslUqfWLVmnXv12vXu1WvXu9a8+RfXmrXrXavXrnevWrPes3rNOs+qNeuwas06WglCUAIMlgC7+k511y+C/xS8b//e4dJ+IeF9+vce3Kd/2OA+d/ce3Kd/78HSfiHhAbzjlVRUlABJgKBvCxUVJUASIBUVFSXALi1AKioqqh6bAKmoqKgoAVJRUVFRAqSioqLqIgmQIAiiO/D/AwDuKUn4ex3n/gAAAABJRU5ErkJggg==);
```
Comment (mejackreed) [inline css/jstree.css]: Awesome! Could we commit these changes separately in a separate `less` file? Then include the hosted css and js . That way we could more easily update in the future if needed.

Comment (dicksonlaw583) [inline css/jstree.css]: Moved to LESS.

**Comment by mejackreed** [comment]
One issue I ran into was when adding a multi collection manifest as a "manifestUri" on instantiation of the Mirador instance.

For example:

```javascript
 myMiradorInstance = Mirador({
         "id": "viewer",
         "layout": "1x1",
         "data": [
           { "manifestUri": "https://digital.library.villanova.edu/Collection/vudl:3/IIIF", "location": 'Villanova' }
         ],
         "windowObjects": [],
         "annotationEndpoint": { "name":"Local Storage", "module": "LocalStorageEndpoint" }
       });
```

Is there a way to mitigate this? It might be difficult for a mirador adopter to have to know what type of manifest it is in advance?

**Comment by mejackreed** [comment]
Oops my bad on this, these work as  `collectionUri` key.

**Comment by aeschylus** [comment]
@dicksonlaw583 Great job with this. I left a series of comments that if we can get cleaned up I can see us merging this to core. One bigger thing that was hard to leave in the comments is that I think this will be easier to maintain if the collection tree and its helper methods are broken out into a separate "view" or module. I realise the jQuery library generates most of the dom for the tree, but it is difficult to follow the logic in the manifestPanel when all the logic of one of its logical children (the tree view) is included in it. The manifestListItems are separate from the manifestPanel, and the treeView has a similar logical distinction from the manifestPanel. Can we make this separation? I think doing so will also encourage less reliance on request state variables in the manifestPanel, as well as transitioning some of the if/else cache checks to event-based reactions.

**Comment by mejackreed** [comment]
@aeschylus are those comments you created posted? I'm not seeing them in this PR?

File: js/src/viewer/manifestListItem.js [hunk 3]
```diff
@@ -45,6 +46,9 @@
       });
 
       this.fetchTplData(this.manifestId);
+      if (this.forcedIndex !== null) {
```
Comment (dicksonlaw583) [inline js/src/viewer/manifestListItem.js]: This is an adaptation I made to allow the collection tree to take over the ordering of manifest items by forcing its "index" number to be a specified value. The current implementation of manifestListItem.js orders by when it was loaded into the global state, which is no longer a good indicator with the way manifests are loaded under collections.

File: js/src/viewer/manifestsPanel.js [hunk 4]
```diff
@@ -48,13 +118,46 @@
...
+          });
+          
+          // When Mirador gets an explicit request to load a collection from a URL
+          _this.eventEmitter.subscribe('ADD_COLLECTION_FROM_URL', function(event, url, source) {
+            _this.addCollectionFromUrl(url, null, true);
+          });
+          
+          // When Mirador gets a request to load an object (manifest or collection) from a URL (e.g. the "Add object from URL" form)
+          _this.eventEmitter.subscribe('ADD_OBJECT_FROM_URL', function(event, url, source) {
+            _this.addObjectFromUrl(url, source, true);
```
Comment (dicksonlaw583) [inline js/src/viewer/manifestsPanel.js]: There are too many hooks into ADD_MANIFEST_FROM_URL for me to safely rework them at this point. This is a compromise until we can re-architect Mirador 3.

File: js/src/viewer/manifestsPanel.js [hunk 5]
```diff
@@ -94,11 +197,18 @@
...
         },
         
+        // Send explicit request for adding a manifest from a URL
         addManifestUrl: function(url) {
           var _this = this;
           _this.eventEmitter.publish('ADD_MANIFEST_FROM_URL', [url, "(Added from URL)"]);
         },
         
+        // Send explicit request for adding a manifest or collection from a URL
+        addObjectUrl: function(url) {
```
Comment (dicksonlaw583) [inline js/src/viewer/manifestsPanel.js]: OK, will do.

File: js/src/viewer/manifestsPanel.js [hunk 6]
```diff
@@ -63,7 +166,7 @@
             // handle interface events
             this.element.find('form#url-load-form').on('submit', function(event) {
               event.preventDefault();
-              _this.addManifestUrl(jQuery(this).find('input').val());
+              _this.addObjectUrl(jQuery(this).find('input').val());
```
Comment (dicksonlaw583) [inline js/src/viewer/manifestsPanel.js]: Same as comment for line 207.

File: js/src/viewer/manifestsPanel.js [hunk 7]
```diff
@@ -129,15 +239,310 @@
            _this.hide();
         },
 
+        // Handler for when manifest data is loaded for the first time
```
Comment (dicksonlaw583) [inline js/src/viewer/manifestsPanel.js]: The side effect is adding a new list item to the manifest list UI. I remember this already being in the code prior to work on the collection tree, this is just added documentation.

File: js/src/viewer/manifestsPanel.js [hunk 8]
```diff
@@ -129,15 +239,310 @@
...
+              resultsWidth: _this.resultsWidth, 
+              state: _this.state,
+              eventEmitter: _this.eventEmitter,
+              forcedIndex: _this.expectedThings.indexOf(newManifest.uri),
+              appendTo: _this.manifestListElement }));
+            _this.element.find('#manifest-search').keyup();
+          }
+        },
+        
+        // Handler for when collection data is loaded for the first time
```
Comment (dicksonlaw583) [inline js/src/viewer/manifestsPanel.js]: I labelled these "first time" because I've only observed them run once when the JSON data is first loaded. On subsequent attempts to load, current Mirador components fetch from the state and bypass the events.

File: js/src/viewer/manifestsPanel.js [hunk 9]
```diff
@@ -129,15 +239,310 @@
...
+        expandNode: function(node) {
+          var _this = this;
+          // Update its children by loading their stated URIs
+          jQuery.each(_this.nodeCollections[node.id], function(_, uri) {
+            _this.updateCollectionFromUrl(uri, node.id);
+          });
+        },
+        
+        // Helper for loading a manifest or collection from a URL
+        addObjectFromUrl: function(url, source) {
```
Comment (dicksonlaw583) [inline js/src/viewer/manifestsPanel.js]: Will fix.

File: js/src/viewer/manifestsPanel.js [hunk 10]
```diff
@@ -129,15 +239,310 @@
...
+          // Cache miss
+          else {
+            // Get the manifest or collection with AJAX
+            jQuery.ajax({
+              url: url,
+              dataType: 'json',
+              type: 'GET',
+              // Fire off the correct event once it finishes loading
+              success: function(data) {
+                switch (data['@type']) {
```
Comment (dicksonlaw583) [inline js/src/viewer/manifestsPanel.js]: Will do. Should I call it publishIiifData(data, url, source) and put it in utilities?

**Comment by dicksonlaw583** [comment]
@aeschylus Sorry for the late replies, I've responded to the comments from your commit and started work on some of the minor refactors.

As for making the manifest panel separate, I plan to discuss moving it out into a configuration (similar to annotation endpoints) at the pre-conference. Would it be OK if I bring a proposal there and port the collection tree to this setup once we've talked this through?

**Comment by dicksonlaw583** [comment]
Sorry for the delay in updating the PR. The collection tree update has passed our internal review and is now ready for merging.

Changes since May (last review):

- A new option named ```manifestsPanel``` has been added to allow replacing the default manifest selection panel.
- The collection tree panel is now a new module named ```CollectionTreeManifestsPanel```, which can be plugged into the option above.

For compatibility with existing installations, I have reset the default back to the classic manifests panel. To change to the collection tree, add this to ```js/src/settings.js``` or to the configuration options passed to ```Mirador()```:

```
manifestsPanel: {
  name: "Collection Tree Manifests Panel",
  module: "CollectionTreeManifestsPanel",
  options: {}
}
```

It should now be possible to pull this in without any visible impact to existing users.
